src/Catalog/Helpers/NonhijackableV2HttpMessageHandlerWrapper.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Services.Metadata.Catalog.Helpers
+{
+    /// <summary>
+    /// A <see cref="HttpMessageHandler"/> that wraps another <see cref="HttpMessageHandler"/> and prevents queries to the V2 feed from being hijacked by the search service.
+    /// </summary>
+    public class NonhijackableV2HttpMessageHandlerWrapper : HttpMessageHandler
+    {
+        /// <param name="inner">The <see cref="HttpMessageHandler"/> to wrap.</param>
+        public NonhijackableV2HttpMessageHandlerWrapper(HttpMessageHandler inner)
+        {
+            _innerHttpMessageHandler = inner ?? throw new ArgumentNullException(nameof(inner));
+        }
+
+        private HttpMessageHandler _innerHttpMessageHandler;
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            request.RequestUri = GetNonhijackableUri(request.RequestUri);
+
+            /// For whatever reason, <see cref="HttpMessageHandler.SendAsync"/> is a protected method.
+            /// As a result, we must use reflection to call <see cref="HttpMessageHandler.SendAsync"/> on <see cref="_innerHttpMessageHandler"/>.
+            var sendAsyncMethod = _innerHttpMessageHandler.GetType().GetMethod(nameof(SendAsync), BindingFlags.NonPublic | BindingFlags.Instance);
+            var sendAsyncResult = (Task<HttpResponseMessage>)sendAsyncMethod.Invoke(_innerHttpMessageHandler, new object[] { request, cancellationToken });
+            return sendAsyncResult;
+        }
+
+        private const string Filter = "$filter=";
+        private const string NonhijackableExpression = "true";
+        private const string And = "%20and%20";
+
+        private const string GetSpecificPackageFormatRegExp = @"Packages\(Id='(.*)',Version='(.*)'\)";
+        private static string GetSpecificPackageNonhijackable = $"Packages?{Filter}{NonhijackableExpression}{And}" + "Id%20eq%20%27{0}%27%20and%20Version%20eq%20%27{1}%27";
+
+        private static IEnumerable<string> HijackableEndpoints = new List<string>
+        {
+            "/Packages",
+            "/Search",
+            "/FindPackagesById"
+        };
+
+        public static Uri GetNonhijackableUri(Uri originalUri)
```

Comment:
>GetNonhijackableUri [](start = 26, length = 19)

This logic does not have to be in an HTTP handler. Please move it out to it's out Utility class (or an existing utility class). Then the HTTP handler can be moved next to its provider.